presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java

```diff
@@ -267,13 +267,44 @@ public PlanWithProperties visitAggregation(AggregationNode node, StreamPreferred
                                 idAllocator.getNextId(),
                                 LOCAL,
                                 child.getNode(),
-                                node.getGroupingKeys(),
+                                groupingKeys,
                                 Optional.empty()),
                         child.getProperties());
                 return rebaseAndDeriveProperties(node, ImmutableList.of(exchange));
             }
 
-            return planAndEnforceChildren(node, requiredProperties, requiredProperties);
+            StreamPreferredProperties childRequirements = parentPreferences
+                    .constrainTo(node.getSource().getOutputSymbols())
+                    .withDefaultParallelism(session)
+                    .withPartitioning(groupingKeys);
+
+            PlanWithProperties child = planAndEnforce(node.getSource(), childRequirements, childRequirements);
+
+            List<Symbol> preGroupedSymbols;
+            if (!groupingKeys.isEmpty()) {
```

Comment:
`groupingKeys` is an union of all symbols in `groupingSets`. As I understand currently there is no point  to find plan with `preGroupedSymbols` matching `groupingKeys` if there are multiple entries in `groupingSets`. Maybe you should collect this only when `groupingSets.size() == 1`?